<div class="alert">  
    <center><h1><strong>Final Project: NLP Fake news detection</strong></h1></center>

<div class="alert alert-block alert-info">  
    <center><h2><strong>Analysis / exploration of the data set</strong></h2></center>

<div class="alert alert-block alert-warning">  
<b>Summary of the section </b> 
<hr>
<ul>
    <li>Analyze the chosen dataset and its properties</li>
</ul>
    
<hr>
<b>Source of Data: </b> 
<hr> 
 <a href="https://www.cs.ucsb.edu/~william/data/liar_dataset">https://www.cs.ucsb.edu/~william/data/liar_dataset</a>
   
</div>

In [ ]:
#Importing the required libraries

import numpy as np
import pandas as pd

## 1. Loading the dataset

In [12]:
# Here we are loading th dataset into a pandas dataframe we can do further analysis on it
df = pd.read_csv("datasets/Liar_Dataset.csv")

# We are also going to look at the first 5 records in it so to have an idead on what we are working with
df.head(5)

[ID].json       label                                          statement  \
0  11972.json        TRUE  Building a wall on the U.S.-Mexico border will...   
1  11685.json       FALSE  Wisconsin is on pace to double the number of l...   
2  11096.json       FALSE  Says John McCain has done nothing to help the ...   
3   5209.json   half-true  Suzanne Bonamici supports a plan that will cut...   
4   9524.json  pants-fire  When asked by a reporter whether hes at the ce...   

                                          subject(s)  \
0                                        immigration   
1                                               jobs   
2                    military,veterans,voting-record   
3  medicare,message-machine-2012,campaign-adverti...   
4  campaign-finance,legal-issues,campaign-adverti...   

                            speaker   speaker's job title state info  \
0                        rick-perry              Governor      Texas   
1                 katrina-shankland  State representative  Wisconsin   
2                      donald-trump       President-Elect   New York   
3                     rob-cornilles            consultant     Oregon   
4  state-democratic-party-wisconsin                   NaN  Wisconsin   

  party affiliation  barely true counts  false counts  half true counts  \
0        republican                  30            30                42   
1          democrat                   2             1                 0   
2        republican                  63           114                51   
3        republican                   1             1                 3   
4          democrat                   5             7                 2   

   mostly true counts  pants on fire counts                         venue  
0                  23                    18               Radio interview  
1                   0                     0             a news conference  
2                  37                    61  comments on ABC's This Week.  
3                   1                     1                  a radio show  
4                   2                     7                   a web video

## 2. Exploring the dataset

In [17]:
# Let us first take a look at what we are dealing with
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12787 entries, 0 to 12786
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   [ID].json             12787 non-null  object
 1   label                 12787 non-null  object
 2   statement             12787 non-null  object
 3   subject(s)            12787 non-null  object
 4   speaker               12787 non-null  object
 5   speaker's job title   9222 non-null   object
 6   state info            10040 non-null  object
 7   party affiliation     12787 non-null  object
 8   barely true counts    12787 non-null  int64 
 9   false counts          12787 non-null  int64 
 10  half true counts      12787 non-null  int64 
 11  mostly true counts    12787 non-null  int64 
 12  pants on fire counts  12787 non-null  int64 
 13  venue                 12658 non-null  object
dtypes: int64(5), object(9)
memory usage: 1.4+ MB


<div class="alert alert-block alert-warning">  
<b>In the above we can point a few things about the dataset</b> 
<hr>
<ul>
    <li>1. It has 14 columns and 12787 rows</li>
    <li>2. It has two datatypes: objects (strings) and int64 (numbers) </li>
    <li>3. The first column - which seems to be some kind of ID - is actually a string. And we can see by looking at the data inside the CSV looks like the filename.</li>
    <li>4. We can also notice here that we have a few nulls, as for some columns (such as) 'speaker's job title ', 'state info' and 'venue' the Non-Null Counts are lower than the number of rows. </li>
</ul>
</div>

In [13]:
# Now take a look at the columns inside the dataset so to have an idea about the features we can use further in our model
df.columns

Index(['[ID].json', 'label', 'statement', 'subject(s)', 'speaker',
       'speaker's job title', 'state info', 'party affiliation',
       'barely true counts', 'false counts', 'half true counts',
       'mostly true counts', 'pants on fire counts', 'venue'],
      dtype='object')

TODO: Understand more about the columns and complete where we have ???
Look into PAPER and fact checking websites (politifact)
<div class="alert alert-block alert-warning">  
<b>Further description by matching the columns with the rows</b> 
<hr>
<ul>
    <li>1. '[ID].json' - As already stated, this seems to be a filename. This may not be an useful feature for us to use so we may delete it in the future.</li>
    <li>2. 'label' -  This column is the classification of the news statement, it can be either TRUE, FALSE, half-true, half-false or pants-fire.</li>
    <li>3. 'statement' - This represents the text in the news article. </li>
    <li>3. 'subject(s)' -   The subjects that the news statement talks about. We will further analyze what are them below. </li>
    <li>3. 'speaker' -  The person who talked/wrote about the news statement.  We will further analyze what are them below. </li>
    <li>4. 'speakers job title' -  The speaker's job.  We will further analyze what are them below. </li>
    <li>5. 'state info' -  ???</li>
    <li>7. 'party affiliation' - The political party the speaker is related to. </li>
    <li>8. 'barely true counts' - The counts of statements that are classified as being barely true. Meaning ??? </li>
    <li>9. 'false counts' - The counts of statements that are classified as being barely false </li>
    <li>10. 'half true counts' -  The counts of statements that are classified as being half true. Meaning ??? </li>
    <li>11. 'mostly true counts' -  The counts of statements that are classified as being mostly true. Meaning ??? </li>
    <li>12. 'pants on fire counts' -  The counts of statements that are classified as being pants on fire. Meaning ??? </li>
    <li>13. 'venue' -  ??? </li>
</ul>
</div>

## 3. Cleaning the data

In [23]:
# We already noticed above that we have some null values. Now let us confirm how many and in which columns
df.isnull().sum()

[ID].json                  0
label                      0
statement                  0
subject(s)                 0
speaker                    0
speaker's job title     3565
state info              2747
party affiliation          0
barely true counts         0
false counts               0
half true counts           0
mostly true counts         0
pants on fire counts       0
venue                    129
dtype: int64

<div class="alert alert-block alert-info">  
    <center><h2><strong>Identification of suitable features and implementation of a suitable feature extractor, e.f.
TfidfVectorizer</strong></h2></center>

<div class="alert alert-block alert-info">  
    <center><h2><strong>Evaluation</strong></h2></center>